src/include/storage/sql_table.h

```diff
@@ -27,6 +31,110 @@ class SqlTable {
     DataTable *data_table;
     BlockLayout layout;
     ColumnMap column_map;
+    InverseColumnMap inverse_column_map;
+    // TODO(John): Add 'default_value_map' (dynamic) for col_oid->default_val
+  };
+
+  /**
+   * Iterator for all the slots, claimed or otherwise, in the data table. This is useful for sequential scans.
+   */
+  class SlotIterator {
+    // TODO(Yashwanth): Slot iterator currently flawed, for a scan on a certain version, it MUST begin on the latest
+    // version it sees
+   public:
+    /**
+     * @return reference to the underlying tuple slot
+     */
+    const TupleSlot &operator*() const { return *current_it_; }
+
+    /**
+     * @return pointer to the underlying tuple slot
+     */
+    const TupleSlot *operator->() const { return &(*current_it_); }
+
+    /**
+     * pre-fix increment.
+     * @return self-reference after the iterator is advanced
+     */
+    SlotIterator &operator++() {
+      TERRIER_ASSERT(!is_end_, "Cannot increment an end iterator");
+      current_it_++;
```

Comment:
Is it possible that a dataTable has no data so that its iterator starts at its end? In this case, you will increase the iterator beyond its boundary and not aware of that, because AdvanceOnEndOfDatatable_() has not yet been called.

Reply:
I'll double check the semantics of the iterator, but I think C++ generally demands that incrementing `end` becomes a no-op.  However, this is an undocumented assumption that we either need to confirm and document or correct.

Reply:
The increment of the end iterator is undefined in STL. They don't say that it must be a no-op.

Reply:
Fixed by making `AdvanceOnEndOfDatatable_()` loop until it finds a non-empty DataTable.